False negative for D417

It appears that ruff v0.0.195 fails to detect missing argument docs from a google-formatted docstring (D417) when a newline is present in the `Returns` block.

```python
# test.py
def f(x, y, z):
    """Do f.

    Args:
        x: the value
        z: A final argument

    Returns: 
        the value
    """
    return x

def g(x, y, z):
    """Do g.

    Args:
        x: the value
        z: A final argument

    Returns: the value
    """
    return x

def h(x, y, z):
    """Do h.

    Args:
        x: the value
        z: A final argument
    """
    return x
```

```
$ ruff --version
ruff 0.0.195

$ ruff --select D --show-source test.py
test.py:14:1: D417 Missing argument description in the docstring: `y`
   |
14 | / def g(x, y, z):
15 | |     """Do g.
16 | |
17 | |     Args:
18 | |         x: the value
19 | |         z: A final argument
20 | |
21 | |     Returns: the value
22 | |     """
23 | |     return x
   | |____________^ D417
   |
test.py:25:1: D417 Missing argument description in the docstring: `y`
   |
25 | / def h(x, y, z):
26 | |     """Do h.
27 | |
28 | |     Args:
29 | |         x: the value
30 | |         z: A final argument
31 | |     """
32 | |     return x
   | |____________^ D417
   |
Found 2 error(s).
```
I'm expecting 3 errors for all three functions, not just g and h.

(I promise I don't just sit around looking for corner-case D417 bugs! But checking for missing docstring args is one of the useful features of pydocstyle which I'm hoping ruff can replace)

--- a/README.md
+++ b/README.md
@@ -1437,6 +1437,18 @@ extend-ignore = [
 ]
 ```
 
+Note that Ruff _also_ supports a [`convention`](#convention) setting:
+
+```toml
+[tool.ruff.pydocstyle]
+convention = "google"
+```
+
+However, this setting is purely used to implement robust detection of Google and NumPy-style
+sections, and thus avoid the [false negatives](https://github.com/PyCQA/pydocstyle/issues/459) seen
+in `pydocstyle`; it does not affect which errors are enabled, which is driven by the `select` and
+`ignore` settings, as described above.
+
 ## Development
 
 Ruff is written in Rust (1.65.0). You'll need to install the [Rust toolchain](https://www.rust-lang.org/tools/install)
@@ -1452,8 +1464,8 @@ For development, we use [nightly Rust](https://rust-lang.github.io/rustup/concep
 
 ```shell
 cargo +nightly fmt
-cargo +nightly clippy
-cargo +nightly test
+cargo +nightly clippy --fix --workspace --all-targets --all-features -- -W clippy::pedantic
+cargo +nightly test --all
 ```
 
 ## Releases
@@ -2668,6 +2680,28 @@ expect that any method decorated by a decorator in this list has no
 [tool.ruff.pep8-naming]
 # Allow a shorthand alias, `@stcmthd`, to trigger static method treatment.
 staticmethod-decorators = ["staticmethod", "stcmthd"]
+```
+
+---
+
+### `pydocstyle`
+
+#### [`convention`](#convention)
+
+Whether to use Google-style or Numpy-style conventions when detecting
+docstring sections. By default, conventions will be inferred from
+the available sections.
+
+**Default value**: `"convention"`
+
+**Type**: `Convention`
+
+**Example usage**:
+
+```toml
+[tool.ruff.pydocstyle]
+# Use Google-style docstrings.
+convention = "google"
 ```
 
 ---

--- a/flake8_to_ruff/src/converter.rs
+++ b/flake8_to_ruff/src/converter.rs
@@ -325,6 +325,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);
@@ -382,6 +383,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);
@@ -439,6 +441,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);
@@ -496,6 +499,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);
@@ -558,6 +562,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);
@@ -654,6 +659,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);
@@ -717,6 +723,7 @@ mod tests {
             isort: None,
             mccabe: None,
             pep8_naming: None,
+            pydocstyle: None,
             pyupgrade: None,
         });
         assert_eq!(actual, expected);

--- a/playground/src/ruff_options.ts
+++ b/playground/src/ruff_options.ts
@@ -222,6 +222,13 @@ export const AVAILABLE_OPTIONS: OptionGroup[] = [
             "type": 'Vec<String>',
         },
     ]},
+    {"name": "pydocstyle", "fields": [
+        {
+            "name": "convention",
+            "default": '"convention"',
+            "type": 'Convention',
+        },
+    ]},
     {"name": "pyupgrade", "fields": [
         {
             "name": "keep-runtime-typing",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -40,3 +40,6 @@ force-wrap-aliases = true
 combine-as-imports = true
 force-single-line = true
 single-line-exclusions = ["os", "logging.handlers"]
+
+[tool.ruff.pydocstyle]
+convention = "google"

--- a/resources/test/fixtures/pydocstyle/D417.py
+++ b/resources/test/fixtures/pydocstyle/D417.py
@@ -1,0 +1,84 @@
+def f(x, y, z):
+    """Do f.
+
+    Args:
+        x: the value
+            with a hanging indent
+
+    Returns:
+        the value
+    """
+    return x
+
+
+def f(x, y, z):
+    """Do f.
+
+    Args:
+        x:
+            The whole thing has a hanging indent.
+
+    Returns:
+        the value
+    """
+    return x
+
+
+def f(x, y, z):
+    """Do f.
+
+    Args:
+        x:
+            The whole thing has a hanging indent.
+
+    Returns: the value
+    """
+    return x
+
+
+def f(x, y, z):
+    """Do f.
+
+    Args:
+        x: the value def
+            ghi
+
+    Returns:
+        the value
+    """
+    return x
+
+
+def f(x, y, z):
+    """Do f.
+
+    Args:
+        x: the value
+        z: A final argument
+
+    Returns:
+        the value
+    """
+    return x
+
+
+def f(x, y, z):
+    """Do g.
+
+    Args:
+        x: the value
+        z: A final argument
+
+    Returns: the value
+    """
+    return x
+
+
+def f(x, y, z):
+    """Do h.
+
+    Args:
+        x: the value
+        z: A final argument
+    """
+    return x

--- a/ruff.schema.json
+++ b/ruff.schema.json
@@ -277,6 +277,17 @@
         }
       }
     },
+    "pydocstyle": {
+      "description": "Options for the `pydocstyle` plugin.",
+      "anyOf": [
+        {
+          "$ref": "#/definitions/Pydocstyle"
+        },
+        {
+          "type": "null"
+        }
+      ]
+    },
     "pyupgrade": {
       "description": "Options for the `pyupgrade` plugin.",
       "anyOf": [
@@ -893,6 +904,13 @@
         "YTT303"
       ]
     },
+    "Convention": {
+      "type": "string",
+      "enum": [
+        "google",
+        "numpy"
+      ]
+    },
     "Flake8AnnotationsOptions": {
       "type": "object",
       "properties": {
@@ -1193,6 +1211,23 @@
           "type": [
             "boolean",
             "null"
+          ]
+        }
+      },
+      "additionalProperties": false
+    },
+    "Pydocstyle": {
+      "type": "object",
+      "properties": {
+        "convention": {
+          "description": "Whether to use Google-style or Numpy-style conventions when detecting docstring sections. By default, conventions will be inferred from the available sections.",
+          "anyOf": [
+            {
+              "$ref": "#/definitions/Convention"
+            },
+            {
+              "type": "null"
+            }
           ]
         }
       },

--- a/src/checkers/ast.rs
+++ b/src/checkers/ast.rs
@@ -3876,7 +3876,11 @@ impl<'a> Checker<'a> {
                     || self.settings.enabled.contains(&CheckCode::D416)
                     || self.settings.enabled.contains(&CheckCode::D417)
                 {
-                    pydocstyle::plugins::sections(self, &docstring);
+                    pydocstyle::plugins::sections(
+                        self,
+                        &docstring,
+                        self.settings.pydocstyle.convention.as_ref(),
+                    );
                 }
             }
         }

--- a/src/pydocstyle/mod.rs
+++ b/src/pydocstyle/mod.rs
@@ -1,5 +1,6 @@
 pub mod helpers;
 pub mod plugins;
+pub mod settings;
 
 #[cfg(test)]
 mod tests {
@@ -11,6 +12,7 @@ mod tests {
 
     use crate::checks::CheckCode;
     use crate::linter::test_path;
+    use crate::pydocstyle::settings::{Convention, Settings};
     use crate::settings;
 
     #[test_case(CheckCode::D100, Path::new("D.py"); "D100")]
@@ -70,6 +72,56 @@ mod tests {
         )?;
         checks.sort_by_key(|check| check.location);
         insta::assert_yaml_snapshot!(snapshot, checks);
+        Ok(())
+    }
+
+    #[test]
+    fn d417_unspecified() -> Result<()> {
+        let mut checks = test_path(
+            Path::new("./resources/test/fixtures/pydocstyle/D417.py"),
+            &settings::Settings {
+                // When inferring the convention, we'll see a few false negatives.
+                // See: https://github.com/PyCQA/pydocstyle/issues/459.
+                pydocstyle: Settings { convention: None },
+                ..settings::Settings::for_rule(CheckCode::D417)
+            },
+        )?;
+        checks.sort_by_key(|check| check.location);
+        insta::assert_yaml_snapshot!(checks);
+        Ok(())
+    }
+
+    #[test]
+    fn d417_google() -> Result<()> {
+        let mut checks = test_path(
+            Path::new("./resources/test/fixtures/pydocstyle/D417.py"),
+            &settings::Settings {
+                // With explicit Google convention, we should flag every function.
+                pydocstyle: Settings {
+                    convention: Some(Convention::Google),
+                },
+                ..settings::Settings::for_rule(CheckCode::D417)
+            },
+        )?;
+        checks.sort_by_key(|check| check.location);
+        insta::assert_yaml_snapshot!(checks);
+        Ok(())
+    }
+
+    #[test]
+    fn d417_numpy() -> Result<()> {
+        let mut checks = test_path(
+            Path::new("./resources/test/fixtures/pydocstyle/D417.py"),
+            &settings::Settings {
+                // With explicit Google convention, we shouldn't flag anything.
+                pydocstyle: Settings {
+                    convention: Some(Convention::Numpy),
+                },
+                ..settings::Settings::for_rule(CheckCode::D417)
+            },
+        )?;
+        checks.sort_by_key(|check| check.location);
+        insta::assert_yaml_snapshot!(checks);
         Ok(())
     }
 }

--- a/src/pydocstyle/plugins.rs
+++ b/src/pydocstyle/plugins.rs
@@ -15,6 +15,7 @@ use crate::docstrings::definition::{Definition, DefinitionKind, Docstring};
 use crate::docstrings::sections::{section_contexts, SectionContext};
 use crate::docstrings::styles::SectionStyle;
 use crate::pydocstyle::helpers::{leading_quote, logical_line};
+use crate::pydocstyle::settings::Convention;
 use crate::visibility::{is_init, is_magic, is_overload, is_override, is_staticmethod, Visibility};
 
 /// D100, D101, D102, D103, D104, D105, D106, D107
@@ -855,7 +856,7 @@ pub fn not_empty(checker: &mut Checker, docstring: &Docstring) -> bool {
 
 /// D212, D214, D215, D405, D406, D407, D408, D409, D410, D411, D412, D413,
 /// D414, D416, D417
-pub fn sections(checker: &mut Checker, docstring: &Docstring) {
+pub fn sections(checker: &mut Checker, docstring: &Docstring, convention: Option<&Convention>) {
     let body = docstring.body;
 
     let lines: Vec<&str> = LinesWithTrailingNewline::from(body).collect();
@@ -863,17 +864,31 @@ pub fn sections(checker: &mut Checker, docstring: &Docstring) {
         return;
     }
 
-    // First, interpret as NumPy-style sections.
-    let mut found_numpy_section = false;
-    for context in &section_contexts(&lines, &SectionStyle::NumPy) {
-        found_numpy_section = true;
-        numpy_section(checker, docstring, context);
-    }
+    match convention {
+        Some(Convention::Google) => {
+            for context in &section_contexts(&lines, &SectionStyle::Google) {
+                google_section(checker, docstring, context);
+            }
+        }
+        Some(Convention::Numpy) => {
+            for context in &section_contexts(&lines, &SectionStyle::NumPy) {
+                numpy_section(checker, docstring, context);
+            }
+        }
+        None => {
+            // First, interpret as NumPy-style sections.
+            let mut found_numpy_section = false;
+            for context in &section_contexts(&lines, &SectionStyle::NumPy) {
+                found_numpy_section = true;
+                numpy_section(checker, docstring, context);
+            }
 
-    // If no such sections were identified, interpret as Google-style sections.
-    if !found_numpy_section {
-        for context in &section_contexts(&lines, &SectionStyle::Google) {
-            google_section(checker, docstring, context);
+            // If no such sections were identified, interpret as Google-style sections.
+            if !found_numpy_section {
+                for context in &section_contexts(&lines, &SectionStyle::Google) {
+                    google_section(checker, docstring, context);
+                }
+            }
         }
     }
 }

--- a/src/pydocstyle/settings.rs
+++ b/src/pydocstyle/settings.rs
@@ -1,0 +1,44 @@
+//! Settings for the `pydocstyle` plugin.
+
+use ruff_macros::ConfigurationOptions;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub enum Convention {
+    Google,
+    Numpy,
+}
+
+#[derive(
+    Debug, PartialEq, Eq, Serialize, Deserialize, Default, ConfigurationOptions, JsonSchema,
+)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case", rename = "Pydocstyle")]
+pub struct Options {
+    #[option(
+        default = r#""convention""#,
+        value_type = "Convention",
+        example = r#"
+            # Use Google-style docstrings.
+            convention = "google"
+        "#
+    )]
+    /// Whether to use Google-style or Numpy-style conventions when detecting
+    /// docstring sections. By default, conventions will be inferred from
+    /// the available sections.
+    pub convention: Option<Convention>,
+}
+
+#[derive(Debug, Default, Hash)]
+pub struct Settings {
+    pub convention: Option<Convention>,
+}
+
+impl Settings {
+    pub fn from_options(options: Options) -> Self {
+        Self {
+            convention: options.convention,
+        }
+    }
+}

--- a/src/settings/configuration.rs
+++ b/src/settings/configuration.rs
@@ -21,7 +21,8 @@ use crate::settings::types::{
 };
 use crate::{
     flake8_annotations, flake8_bugbear, flake8_errmsg, flake8_import_conventions, flake8_quotes,
-    flake8_tidy_imports, flake8_unused_arguments, fs, isort, mccabe, pep8_naming, pyupgrade,
+    flake8_tidy_imports, flake8_unused_arguments, fs, isort, mccabe, pep8_naming, pydocstyle,
+    pyupgrade,
 };
 
 #[derive(Debug, Default)]
@@ -62,6 +63,7 @@ pub struct Configuration {
     pub isort: Option<isort::settings::Options>,
     pub mccabe: Option<mccabe::settings::Options>,
     pub pep8_naming: Option<pep8_naming::settings::Options>,
+    pub pydocstyle: Option<pydocstyle::settings::Options>,
     pub pyupgrade: Option<pyupgrade::settings::Options>,
 }
 
@@ -156,6 +158,7 @@ impl Configuration {
             isort: options.isort,
             mccabe: options.mccabe,
             pep8_naming: options.pep8_naming,
+            pydocstyle: options.pydocstyle,
             pyupgrade: options.pyupgrade,
         })
     }
@@ -217,6 +220,7 @@ impl Configuration {
             isort: self.isort.or(config.isort),
             mccabe: self.mccabe.or(config.mccabe),
             pep8_naming: self.pep8_naming.or(config.pep8_naming),
+            pydocstyle: self.pydocstyle.or(config.pydocstyle),
             pyupgrade: self.pyupgrade.or(config.pyupgrade),
         }
     }

--- a/src/settings/mod.rs
+++ b/src/settings/mod.rs
@@ -22,7 +22,8 @@ use crate::settings::types::{
 };
 use crate::{
     flake8_annotations, flake8_bugbear, flake8_errmsg, flake8_import_conventions, flake8_quotes,
-    flake8_tidy_imports, flake8_unused_arguments, isort, mccabe, pep8_naming, pyupgrade,
+    flake8_tidy_imports, flake8_unused_arguments, isort, mccabe, pep8_naming, pydocstyle,
+    pyupgrade,
 };
 
 pub mod configuration;
@@ -68,6 +69,7 @@ pub struct Settings {
     pub isort: isort::settings::Settings,
     pub mccabe: mccabe::settings::Settings,
     pub pep8_naming: pep8_naming::settings::Settings,
+    pub pydocstyle: pydocstyle::settings::Settings,
     pub pyupgrade: pyupgrade::settings::Settings,
 }
 
@@ -193,6 +195,10 @@ impl Settings {
                 .pep8_naming
                 .map(pep8_naming::settings::Settings::from_options)
                 .unwrap_or_default(),
+            pydocstyle: config
+                .pydocstyle
+                .map(pydocstyle::settings::Settings::from_options)
+                .unwrap_or_default(),
             pyupgrade: config
                 .pyupgrade
                 .as_ref()
@@ -233,6 +239,7 @@ impl Settings {
             isort: isort::settings::Settings::default(),
             mccabe: mccabe::settings::Settings::default(),
             pep8_naming: pep8_naming::settings::Settings::default(),
+            pydocstyle: pydocstyle::settings::Settings::default(),
             pyupgrade: pyupgrade::settings::Settings::default(),
         }
     }
@@ -269,6 +276,7 @@ impl Settings {
             isort: isort::settings::Settings::default(),
             mccabe: mccabe::settings::Settings::default(),
             pep8_naming: pep8_naming::settings::Settings::default(),
+            pydocstyle: pydocstyle::settings::Settings::default(),
             pyupgrade: pyupgrade::settings::Settings::default(),
         }
     }
@@ -326,6 +334,7 @@ impl Hash for Settings {
         self.isort.hash(state);
         self.mccabe.hash(state);
         self.pep8_naming.hash(state);
+        self.pydocstyle.hash(state);
         self.pyupgrade.hash(state);
     }
 }

--- a/src/settings/options.rs
+++ b/src/settings/options.rs
@@ -9,7 +9,8 @@ use crate::checks_gen::CheckCodePrefix;
 use crate::settings::types::{PythonVersion, SerializationFormat, Version};
 use crate::{
     flake8_annotations, flake8_bugbear, flake8_errmsg, flake8_import_conventions, flake8_quotes,
-    flake8_tidy_imports, flake8_unused_arguments, isort, mccabe, pep8_naming, pyupgrade,
+    flake8_tidy_imports, flake8_unused_arguments, isort, mccabe, pep8_naming, pydocstyle,
+    pyupgrade,
 };
 
 #[derive(
@@ -368,6 +369,9 @@ pub struct Options {
     #[option_group]
     /// Options for the `pep8-naming` plugin.
     pub pep8_naming: Option<pep8_naming::settings::Options>,
+    #[option_group]
+    /// Options for the `pydocstyle` plugin.
+    pub pydocstyle: Option<pydocstyle::settings::Options>,
     #[option_group]
     /// Options for the `pyupgrade` plugin.
     pub pyupgrade: Option<pyupgrade::settings::Options>,

--- a/src/settings/pyproject.rs
+++ b/src/settings/pyproject.rs
@@ -198,6 +198,7 @@ mod tests {
                     isort: None,
                     mccabe: None,
                     pep8_naming: None,
+                    pydocstyle: None,
                     pyupgrade: None,
                 })
             })
@@ -249,6 +250,7 @@ line-length = 79
                     isort: None,
                     mccabe: None,
                     pep8_naming: None,
+                    pydocstyle: None,
                     pyupgrade: None,
                 })
             })
@@ -300,6 +302,7 @@ exclude = ["foo.py"]
                     isort: None,
                     mccabe: None,
                     pep8_naming: None,
+                    pydocstyle: None,
                     pyupgrade: None,
                 })
             })
@@ -351,6 +354,7 @@ select = ["E501"]
                     isort: None,
                     mccabe: None,
                     pep8_naming: None,
+                    pydocstyle: None,
                     pyupgrade: None,
                 })
             })
@@ -403,6 +407,7 @@ ignore = ["E501"]
                     isort: None,
                     mccabe: None,
                     pep8_naming: None,
+                    pydocstyle: None,
                     pyupgrade: None,
                 })
             })
@@ -541,6 +546,7 @@ other-attribute = 1
                     ]),
                     staticmethod_decorators: Some(vec!["staticmethod".to_string()]),
                 }),
+                pydocstyle: None,
                 pyupgrade: None,
             }
         );

